File: src/lib/avadhan/orthogonalizer.ts
// Avadhan Orthogonalizer
// Implements Gram-Schmidt orthogonalization for interference control
// Reference: Whitepaper Section 5 & 13

import { Slot } from './types'

/**
 * Compute the inner product (dot product) of two vectors
 */
export function innerProduct(a: number[], b: number[]): number {
    if (a.length !== b.length) {
        throw new Error('Vectors must have same dimension')
    }
    return a.reduce((sum, val, i) => sum + val * b[i], 0)
}

/**
 * Compute the L2 norm of a vector
 */
export function norm(v: number[]): number {
    return Math.sqrt(innerProduct(v, v))
}

/**
 * Normalize a vector to unit length
 */
export function normalize(v: number[], eps: number = 1e-6): number[] {
    const n = norm(v)
    if (n < eps) return v.map(() => 0)
    return v.map(x => x / n)
}

/**
 * Subtract vector b from a (element-wise)
 */
export function subtract(a: number[], b: number[]): number[] {
    return a.map((val, i) => val - b[i])
}

/**
 * Scale vector by scalar
 */
export function scale(v: number[], s: number): number[] {
    return v.map(x => x * s)
}

/**
 * Soft Gram-Schmidt Orthogonalization
 * Enforces orthogonality across all slot state vectors
 * 
 * From whitepaper:
 * S̃_i = S_i - Σ_{j<i} (⟨S_i, S_j⟩ / |S_j|²) S_j
 * S'_i = S̃_i / |S̃_i|
 */
export function orthogonalize(slots: Slot[], eps: number = 1e-6): Slot[] {
    if (slots.length === 0) return []

    const orthSlots: Slot[] = []
    const orthVectors: number[][] = []

    for (let i = 0; i < slots.length; i++) {
        let v = [...slots[i].stateVector]

        // Subtract projections onto all previous orthogonal vectors
        for (let j = 0; j < i; j++) {
            const u = orthVectors[j]
            const uNormSq = innerProduct(u, u)
            if (uNormSq > eps) {
                const coeff = innerProduct(v, u) / uNormSq
                const projection = scale(u, coeff)
                v = subtract(v, projection)
            }
        }

        // Normalize
        const vNorm = norm(v)
        if (vNorm > eps) {
            v = scale(v, 1 / vNorm)
        }

        orthVectors.push(v)
        orthSlots.push({
            ...slots[i],
            stateVector: v,
        })
    }

    return orthSlots
}

/**
 * Compute orthogonality matrix: ⟨S_i, S_j⟩ for all pairs
 * Ideal: all off-diagonal elements should be 0
 */
export function computeOrthogonalityMatrix(slots: Slot[]): number[][] {
    const n = slots.length
    const matrix: number[][] = Array(n).fill(null).map(() => Array(n).fill(0))

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            matrix[i][j] = innerProduct(
                normalize(slots[i].stateVector),
                normalize(slots[j].stateVector)
            )
        }
    }

    return matrix
}

/**
 * Compute interference score
 * Lower is better; 0 means perfect orthogonality
 */
export function computeInterferenceScore(slots: Slot[]): number {
    const matrix = computeOrthogonalityMatrix(slots)
    let totalInterference = 0
    let count = 0

    for (let i = 0; i < matrix.length; i++) {
        for (let j = 0; j < matrix.length; j++) {
            if (i !== j) {
                totalInterference += Math.abs(matrix[i][j])
                count++
            }
        }
    }

    return count > 0 ? totalInterference / count : 0
}

/**
 * Compute orthogonality loss for training
 * L_orth = λ Σ_{i≠j} |⟨S_i, S_j⟩|²
 */
export function computeOrthogonalityLoss(
    slots: Slot[],
    lambda: number = 0.1
): number {
    let loss = 0

    for (let i = 0; i < slots.length; i++) {
        for (let j = 0; j < slots.length; j++) {
            if (i !== j) {
                const ip = innerProduct(
                    normalize(slots[i].stateVector),
                    normalize(slots[j].stateVector)
                )
                loss += ip * ip
            }
        }
    }

    return lambda * loss
}

/**
 * Apply repulsion force to reduce interference
 * From whitepaper: Ṡ_i = -4 Σ_{j≠i} (Q_j Q_j^T) S_i
 */
export function applyRepulsion(
    slots: Slot[],
    learningRate: number = 0.01
): Slot[] {
    const normalizedVectors = slots.map(s => normalize(s.stateVector))

    return slots.map((slot, i) => {
        let repulsionForce = new Array(slot.stateVector.length).fill(0)

        for (let j = 0; j < slots.length; j++) {
            if (i !== j) {
                const qj = normalizedVectors[j]
                // Compute Q_j Q_j^T S_i (outer product applied to S_i)
                const dotProduct = innerProduct(qj, slot.stateVector)
                const contribution = scale(qj, dotProduct)
                repulsionForce = repulsionForce.map((v, k) => v + contribution[k])
            }
        }

        // Apply repulsion: S_i = S_i - 4 * lr * repulsionForce
        const newVector = slot.stateVector.map(
            (v, k) => v - 4 * learningRate * repulsionForce[k]
        )

        return {
            ...slot,
            stateVector: normalize(newVector),
        }
    })
}

File: backend/avadhan/orthogonalizer.py
"""
Avadhan Orthogonalizer - GPU-accelerated Gram-Schmidt
Implements interference control between attention slots
"""
import torch
import torch.nn.functional as F
from typing import List, Dict


class Orthogonalizer:
    """
    GPU-accelerated orthogonalization for slot state vectors
    Implements Gram-Schmidt with batched operations
    
    From whitepaper:
    S̃_i = S_i - Σ_{j<i} (⟨S_i, S_j⟩ / |S_j|²) S_j
    S'_i = S̃_i / |S̃_i|
    """
    
    def __init__(self, device: torch.device = None, eps: float = 1e-6):
        self.device = device or torch.device("cpu")
        self.eps = eps
    
    def orthogonalize(self, slots: List[Dict]) -> List[Dict]:
        """
        Apply Gram-Schmidt orthogonalization to all slots
        
        Args:
            slots: List of slot dicts with 'vector' key
        
        Returns:
            Slots with orthogonalized vectors
        """
        if len(slots) < 2:
            return slots
        
        # Extract and stack vectors
        vectors = torch.stack([s["vector"] for s in slots])
        
        # Apply batched Gram-Schmidt
        orth_vectors = self._gram_schmidt_batched(vectors)
        
        # Update slots with orthogonalized vectors
        for i, slot in enumerate(slots):
            slot["vector"] = orth_vectors[i]
        
        return slots
    
    def _gram_schmidt_batched(self, vectors: torch.Tensor) -> torch.Tensor:
        """
        Batched Gram-Schmidt orthogonalization
        
        Args:
            vectors: [N, D] tensor of slot vectors
        
        Returns:
            [N, D] tensor of orthogonalized vectors
        """
        n, d = vectors.shape
        device = vectors.device
        
        # Normalize input
        vectors = F.normalize(vectors, dim=1, eps=self.eps)
        
        # Orthogonalize iteratively
        orth = torch.zeros_like(vectors)
        
        for i in range(n):
            v = vectors[i].clone()
            
            # Subtract projections onto all previous orthogonal vectors
            if i > 0:
                # Compute all projections at once
                U = orth[:i]  # [i, D]
                coeffs = torch.mv(U, v)  # [i] = ⟨v, u_j⟩
                projection = torch.sum(coeffs.unsqueeze(1) * U, dim=0)  # [D]
                v = v - projection
            
            # Normalize
            norm = v.norm()
            if norm > self.eps:
                v = v / norm
            
            orth[i] = v
        
        return orth
    
    def compute_matrix(self, slots: List[Dict]) -> torch.Tensor:
        """
        Compute orthogonality matrix: ⟨S_i, S_j⟩ for all pairs
        Ideal: all off-diagonal elements should be 0
        
        Returns:
            [N, N] similarity matrix
        """
        if len(slots) < 2:
            return torch.eye(len(slots), device=self.device)
        
        vectors = torch.stack([s["vector"] for s in slots])
        vectors = F.normalize(vectors, dim=1, eps=self.eps)
        
        # Compute similarity matrix
        sim_matrix = torch.mm(vectors, vectors.t())
        
        return sim_matrix
    
    def compute_interference_rate(self, slots: List[Dict]) -> float:
        """
        Compute interference rate (average off-diagonal similarity)
        Lower is better; 0 means perfect orthogonality
        """
        if len(slots) < 2:
            return 0.0
        
        matrix = self.compute_matrix(slots)
        n = matrix.shape[0]
        
        # Get off-diagonal elements
        mask = ~torch.eye(n, dtype=torch.bool, device=self.device)
        off_diagonal = matrix[mask]
        
        # Average absolute value
        interference = off_diagonal.abs().mean().item()
        
        return interference
    
    def compute_loss(
        self, 
        slots: List[Dict], 
        weight: float = 0.1
    ) -> torch.Tensor:
        """
        Compute orthogonality loss for training
        L_orth = λ Σ_{i≠j} |⟨S_i, S_j⟩|²
        
        Args:
            slots: List of slot dicts
            weight: Orthogonality weight (λ/β)
        
        Returns:
            Scalar loss tensor
        """
        if len(slots) < 2:
            return torch.tensor(0.0, device=self.device)
        
        vectors = torch.stack([s["vector"] for s in slots])
        vectors = F.normalize(vectors, dim=1, eps=self.eps)
        
        # Compute similarity matrix
        sim_matrix = torch.mm(vectors, vectors.t())
        
        # Mask diagonal (self-similarity)
        n = sim_matrix.shape[0]
        mask = ~torch.eye(n, dtype=torch.bool, device=self.device)
        
        # Sum of squared off-diagonal elements
        off_diagonal = sim_matrix[mask]
        loss = (off_diagonal ** 2).sum()
        
        return weight * loss
    
    def apply_repulsion(
        self, 
        slots: List[Dict], 
        learning_rate: float = 0.01
    ) -> List[Dict]:
        """
        Apply repulsion force to reduce interference
        From whitepaper: Ṡ_i = -4 Σ_{j≠i} (Q_j Q_j^T) S_i
        """
        if len(slots) < 2:
            return slots
        
        vectors = torch.stack([s["vector"] for s in slots])
        vectors = F.normalize(vectors, dim=1, eps=self.eps)
        
        n = vectors.shape[0]
        new_vectors = vectors.clone()
        
        for i in range(n):
            repulsion = torch.zeros_like(vectors[i])
            
            for j in range(n):
                if i != j:
                    # Outer product applied: Q_j (Q_j^T S_i)
                    dot = torch.dot(vectors[j], vectors[i])
                    repulsion += dot * vectors[j]
            
            # Apply repulsion
            new_vectors[i] = vectors[i] - 4 * learning_rate * repulsion
            new_vectors[i] = F.normalize(new_vectors[i], dim=0, eps=self.eps)
        
        # Update slots
        for i, slot in enumerate(slots):
            slot["vector"] = new_vectors[i]
        
        return slots
